[SDAG] Teach FoldConstantArithmetic to match splats inserted into vectors

MacDue opened the pull request:
This teaches FoldConstantArithmetic to match `insert_subvector undef, (splat X), N2` as a splat of X. This pattern can occur for scalable vectors when a fixed-length splat is inserted into an undef vector.

This allows the cases in `fixed-subvector-insert-into-scalable.ll` to constant-folded (where previously they would all be computed at runtime).

MacDue committed:
Precommit tests

RKSimon commented on llvm/test/CodeGen/X86/pr35443.ll:
```diff
@@ -8,7 +8,7 @@
 ; CHECK-LABEL: pr35443:
 ; CHECK:       # %bb.0: # %entry
-; CHECK-NEXT:    vpbroadcastb ac+4(%rip), %xmm0
+; CHECK-NEXT:    vpbroadcastb ac+4(%rip), %ymm0
```
This is unfortunate - SimplifyDemandedVectorElts should narrow this to xmm

MacDue replied:
I may move this fold to AArch64 only, it looks like RISCV may also have some trouble with this form, so I'm not sure applying it generally is worthwhile.

MacDue committed:
[SDAG] Teach FoldConstantArithmetic to match splats inserted into vectors

This teaches FoldConstantArithmetic to match `insert_subvector undef, (splat X), N2`
as a splat of X. This pattern can occur for scalable vectors when a
fixed-length splat is inserted into an undef vector.

This allows the cases in `fixed-subvector-insert-into-scalable.ll` to be
constant-folded (where previously they would all be computed at runtime).

MacDue committed:
Format

MacDue committed:
undef > poison

RKSimon commented:
> (CI failure looks unrelated)

The aarch64 CI has been erratic for a few days now
